osu.Game.Rulesets.Osu/Utils/OsuHitObjectGenerationUtils.cs

```diff
@@ -186,5 +187,37 @@ private static Vector2 rotateVector(Vector2 vector, float rotation)
                 length * MathF.Sin(angle)
             );
         }
+
+        public static bool IsHitObjectOnBeat(OsuBeatmap beatmap, OsuHitObject hitObject, bool downbeatsOnly = false)
+        {
+            var timingPoints = beatmap.ControlPointInfo.TimingPoints;
+            var currentTimingPoint = timingPoints.Reverse().FirstOrDefault(p => p.Time <= hitObject.StartTime);
+
+            if (currentTimingPoint == null)
+                return false;
+
+            double timeSinceTimingPoint = hitObject.StartTime - currentTimingPoint.Time;
+
+            double length = downbeatsOnly
+                ? currentTimingPoint.BeatLength * currentTimingPoint.TimeSignature.Numerator
+                : currentTimingPoint.BeatLength;
+
+            return (timeSinceTimingPoint + 1) % length < 2;
+        }
+
+        public static float GetRelativeTargetAngle(float targetDistance, float offset, bool flowDirection)
```

Comment:
`offset` and `flowDirection` need xmldoc. Placement of method is also questionable. I don't imagine it ever getting used outside of random mod, seems too specific.

Reply:
maybe spacing adjust? its had some of the same issues

Reply:
I don't think that spacing adjust will need this, so I moved it to `OsuModRandom.cs`. It can still be moved back at a later point if needed.